libsolidity/experimental/analysis/FunctionCallGraph.cpp

```diff
@@ -0,0 +1,87 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+// SPDX-License-Identifier: GPL-3.0
+
+#include <libsolidity/experimental/analysis/Analysis.h>
+#include <libsolidity/experimental/analysis/FunctionCallGraph.h>
+
+using namespace solidity::frontend::experimental;
+using namespace solidity::util;
+
+FunctionCallGraph::FunctionCallGraph(solidity::frontend::experimental::Analysis& _analysis):
+	m_analysis(_analysis),
+	m_errorReporter(_analysis.errorReporter()),
+	m_currentNode(nullptr),
+	m_inFunctionDefinition(false)
+{
+}
+
+bool FunctionCallGraph::analyze(SourceUnit const& _sourceUnit)
+{
+	_sourceUnit.accept(*this);
+	// TODO remove debug output before merge
+	//std::cout << annotation().functionCallGraph << std::endl;
+	return !m_errorReporter.hasErrors();
+}
+
+bool FunctionCallGraph::visit(FunctionDefinition const& _functionDefinition)
+{
+	m_inFunctionDefinition = true;
```

Comment:
Should probably assert that ``!m_inFunctionDefinition`` (or after https://github.com/ethereum/solidity/pull/14575/files#r1342650107, ``m_currentFunction == nullptr``) before-hand.